Add basic survey model and related functions

zfrank7777 opened the pull request:
Build basic structure for survey.

Add models: Survey, Question
Add db operations: SurveyFunctions
Add APIs: create_survey, show_survey, show_questions

zfrank7777 committed:
Solve conflict with develop

amjltc295 commented on src/surfing_penguin/models.py:
```diff
@@ -28,3 +29,26 @@ def __repr__(self):
+    title = Column(String(128))
+    content = Column(String(1024))
+    survey_id = Column(Integer, ForeignKey('survey.id'))
+    # survey = relationship("Survey", back_populates="question")
```
Do not push comment unless it will be uncommented under some circumstances. If that is the case, you have to explain ho to use it.

amjltc295 commented on src/surfing_penguin/db_interface.py:
```diff
@@ -44,24 +44,31 @@ def load_user(id):
         self.iter = 0
-        self.questions = []
+        self.name = name
+        survey = Survey(name)
```
Why create a Survey here? It seems to me that SurveyFunctions is not for creating a survey but for doing survey related functions.

amjltc295 replied:
If db_interface would be more complicated in the future, you could break it into several files under db_interface folder. SurveyFunctions doesn't have to be a class; it could be a file with different operations to the database.

amjltc295 replied:
@zfrank7777  Are you going to do this in this PR?

amjltc295 reviewed (changes requested):
Well done. Only some structures need to be modified.

zfrank7777 committed:
Revise SurveyFunctions so that it won't work as an object.

amjltc295 commented on src/surfing_penguin/db_interface.py:
```diff
@@ -60,15 +59,14 @@ def get_all_questions(name):
-        question.survey_id = s.id
+        survey.question_num += 1
+        question.idx = survey.question_num
+        question.survey_id = survey.id
```
These setting should be wrapped into a function, like set_question(title, content, idx, survey_id) or in \_\_init\_\_().

zfrank7777 replied:
I think there should not be __init__() because the SurveyFunctions object just works as a class of functions. And I don't really understand what's the difference to wrapped it into another function.

amjltc295 replied:
I mean it should be in \_\_init\_\_() or set_question() of the __Question__ class, not SurveyFunctions. Setting Question's attributes one by one in new_question() is a bit redundant. It's not wrong but it could be simpler for readability.

amjltc295 commented on src/surfing_penguin/db_interface.py:
```diff
@@ -44,24 +44,29 @@ def load_user(id):
+        question.idx = survey.question_num
+        question.survey_id = survey.id
+        session.add(question)
+        session.add(survey)
```
Why is session.add(survey) here?

zfrank7777 replied:
Because question_num is changed.

amjltc295 replied:
@zfrank7777  I think session would do that for you? I tried deleting that line and it makes no difference.

amjltc295 commented on src/surfing_penguin/db_interface.py:
```diff
@@ -44,24 +44,29 @@ def load_user(id):
-        question = self.get(id)
-        self.question.remove(question)
+    def new_question(survey, data):
+        # data should be a dictionary, with key "content" and "title"
```
It would be better to explain the inputs in standard format if it's complicated. It could be used to generate document automatically by Sphinx (http://www.sphinx-doc.org/en/master/contents.html) as well.
For example:
```
"""
(Explain the function here)

Args:
      survey: Survey object in ORM
     data: a dictionary with key 'content' and 'title'
Returns:
    (explain if there is one)
"""
```
Also questions should include other attributes later. Please add a TODO for this.

amjltc295 replied:
Okay, I saw the TODO in model.py

amjltc295 commented on src/surfing_penguin/routes.py:
```diff
@@ -1,34 +1,83 @@
+            'surveyname': api.payload['surveyname'],
+            'questions': q
+        }
+
```
Don't use a single character for variables unless it's in a simple loop.

amjltc295 reviewed (changes requested):
Is Qstnr class outdated now?

zfrank7777 committed:
Modify Question model (add init function).

Revise variable name in route.py.
Move the assigning value command for a question to the init function of Question model.

zfrank7777 commented:
Yes, there's no Qstnr anymore. It's now called Survey.

amjltc295 commented on src/surfing_penguin/db_interface.py:
```diff
@@ -60,13 +60,15 @@ def get_all_questions(name):
-        question.content = data["content"]
+        """
+        add a question to a survey
+
```
Please capitalize the first word ...

amjltc295 reviewed (changes requested):
The for loop in create_survey() should be put in new_survey() since it's not for I/O processing. Also, there is no difference if two surveys with the same name are created. Survey ID should be given when a new survey is created.
On the other hand, I think most attributes in a model should be in the api_model as well (e.g. question_id in Question, question_num in Survey).

zfrank7777 committed:
Modify new_survey so that it includes new_questions functions.

Add id, idx in api model survey, question, respectively.
Replace show_questions with show_survey,
    which works as a way to pick up a certain survey.
Add id-searching for survey.

zfrank7777 committed:
Add some comments.

zfrank7777 committed:
Create directory: db_interface.

Replace db_interface.py with db_interface directory.
Divide the content into two .py files.
Break id/name searching for survey into two functions/apis.

zfrank7777 committed:
Revise comment

Commit:
Add TODOs

amjltc295 commented:
I added some TODOs for the back end. It looks good to me for this PR.
